### ISSUE TITLE
Cannot add methods to abstract type - moving to 0.5, Julia issue 14919

### ISSUE THREAD

**Issue by tawheeler** [sisl/BayesNets.jl#28]
It looks like `Base.call` is no longer supported for abstract types in Julia 0.5. See [issue 14919](https://github.com/JuliaLang/julia/issues/14919).

At the time of this writing we use call for CPDs to allow you to condition on them. ie, `cpd(assignemnt) -> pdf`.  In 0.5 we may have to create a macro that does this for you that you call after defining your cpd, or stop using `call` and define a function with a name, maybe `condition`.

`condition(cpd, assignment)` or similar.

**Comment by mykelk** [comment]
You can use call for concrete types, right? I think we currently have something like this:
```julia
Base.call(cpd::CPD) = call(cpd, Assignment())
```
but won't it work to do something like this?
```julia
Base.call{T <: CPD}(cpd::T) = call(cpd, Assignment())
```

**Comment by tawheeler** [comment]
Good idea.
It is worth noting that what was once
`Base.call{T<:Bar}(foo::Foo{T}, args) = stuff`
is now
`(foo::Foo{T}){T<:Bar}(args) = stuff
 in 0.5

For `abstract Foo` we have:
`(foo::Foo)(x::Int) = x + 2` yields _ERROR: cannot add methods to an abstract type_
`(foo::T){T}(x::Int) = x+2` yields _ERROR: function type in method definition is not a type_

The good news is that we are already defining `Base.call(cpd, assignment)` for each concrete type. We currently are failing on 
```
(cpd::T){T<:CPD}() = (cpd)(Assignment()) # cpd()
(cpd::T){T<:CPD}(pair::Pair{NodeName}...) = (cpd)(Assignment(pair)) # cpd(:A=>1)
```

**Comment by mykelk** [comment]
Hmmm... So, should we define those for all the concrete CPDs or make a macro?

For the change in call syntax, it would be nice to support both 0.4 and 0.5 (for now at least).

**Comment by cortner** [comment]
In case you are interested, this macro works for me (so far)
```julia
macro pot(fsig)
   @assert fsig.head == :type
   tname, tparams = t_info(fsig.args[2])
   tname = esc(tname)
   for n = 1:length(tparams)
      tparams[n] = esc(tparams[n])
   end
   sym = esc(:x)
   return quote
      $(esc(fsig))
      ($sym::$tname){$(tparams...)}(args...) = evaluate($sym, args...)
   end
end


# t_info extracts type name as symbol and type parameters as an array
t_info(ex::Symbol) = (ex, tuple())
t_info(ex::Expr) = ex.head == :(<:) ? t_info(ex.args[1]) : (ex, ex.args[2:end])
```

Then I define new types via 
```julia
@pot type A end
```

Problem is I can't add documentation to the types.

**Comment by tawheeler** [comment]
Thank you @cortner. 
For documentation we could perhaps do:
```
"""
my docs
"""
type A <: CPD end
@cpd A
```

**Comment by cortner** [comment]
yes, I should adopt that as well; I didn't like it visually, but hopefully it will be just a temporary workaround.

**Comment by tawheeler** [comment]
Fixed with 6508180
I ended up going with this:

```
macro define_call(cpd_type)
    return quote
        @compat (cpd::$cpd_type)() = (cpd)(Assignment()) # cpd()
        @compat (cpd::$cpd_type)(pair::Pair{NodeName}...) = (cpd)(Assignment(pair)) # cpd(:A=>1)
    end
end
```

So now for a new CPD you call:

```
type MyCPD
   ... stuff ...
end

@define_call MyCPD
@compat function (cpd::MyCPD)(a::Assignment)
   ... my cpd code ...
end
```
